incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -2516,7 +2706,7 @@ protected void handleKafkaData(
                 kafkaDataEx != null && kafkaDataEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaDataEx.merged() : null;
             final KafkaKeyFW key = kafkaMergedDataEx != null ? kafkaMergedDataEx.key() : null;
 
-            if (key != null)
+            if (key != null && payload != null)
             {
                 if (key.length() == (delegate.clientId.length()))
                 {
```

Comment:
Lets pull out `int keyLen = key.length()` once instead of calling it multiple times in the `if-else-if`.